libkbfs/folder_branch_ops.go

```diff
@@ -1024,6 +1037,296 @@ func (fbo *folderBranchOps) kickOffPartialSyncIfNeeded(
 	fbo.kickOffPartialSync(ctx, lState, syncConfig, rmd)
 }
 
+func (fbo *folderBranchOps) markRecursive(
+	ctx context.Context, lState *lockState, node Node,
+	rmd ImmutableRootMetadata, tag string, cacheType DiskBlockCacheType) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	err := fbo.blocks.MarkNode(ctx, lState, node, rmd, tag, cacheType)
+	if err != nil {
+		return err
+	}
+
+	if node.EntryType() != Dir {
+		return nil
+	}
+
+	p := fbo.nodeCache.PathFromNode(node)
+	children, err := fbo.blocks.GetChildren(ctx, lState, rmd, p)
+	if err != nil {
+		return err
+	}
+	for child := range children {
+		childNode, _, err := fbo.Lookup(ctx, node, child)
+		if err != nil {
+			return err
+		}
+		err = fbo.markRecursive(ctx, lState, childNode, rmd, tag, cacheType)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// doPartialMarkAndSweep runs a mark-and-sweep algorithm against all
+// the currently-synced paths, to delete any blocks not reachable from
+// one of these paths.
+func (fbo *folderBranchOps) doPartialMarkAndSweep(
+	ctx context.Context, syncConfig keybase1.FolderSyncConfig,
+	latestMerged ImmutableRootMetadata) (err error) {
+	fbo.log.CDebugf(
+		ctx, "Starting partial mark-and-sweep at revision %d",
+		latestMerged.Revision())
+	defer func() {
+		fbo.deferLog.CDebugf(ctx, "Partial mark-and-sweep done: %+v", err)
+	}()
+
+	if syncConfig.Mode != keybase1.FolderSyncMode_PARTIAL {
+		return errors.Errorf(
+			"Bad mode passed to partial unsync: %+v", syncConfig.Mode)
+	} else if len(syncConfig.Paths) == 0 {
+		return nil
+	}
+
+	rootNode, _, _, err := fbo.getRootNode(ctx)
+	if err != nil {
+		return err
+	}
+	tag := ctx.Value(CtxFBOIDKey).(string)
+	lState := makeFBOLockState()
+	cacheType := DiskBlockSyncCache
+	err = fbo.blocks.MarkNode(
+		ctx, lState, rootNode, latestMerged, tag, cacheType)
+	if err != nil {
+		return err
+	}
+
+pathLoop:
+	for _, p := range syncConfig.Paths {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		fbo.log.CDebugf(ctx, "Marking %s", p)
+
+		// Mark the parent directories.
+		parentPath, syncedElem := stdpath.Split(p)
+		parents := strings.Split(strings.TrimSuffix(parentPath, "/"), "/")
+		currNode := rootNode
+		for _, parent := range parents {
+			if len(parent) == 0 {
+				continue
+			}
+			// TODO: parallelize the parent fetches and lookups.
+			currNode, _, err = fbo.Lookup(ctx, currNode, parent)
+			switch errors.Cause(err).(type) {
+			case NoSuchNameError:
+				fbo.log.CDebugf(ctx, "Synced path %s doesn't exist yet", p)
+				continue pathLoop
+			case nil:
+			default:
+				return err
+			}
+
+			err = fbo.blocks.MarkNode(
+				ctx, lState, currNode, latestMerged, tag, cacheType)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Now mark everything rooted at this path.
+		currNode, _, err = fbo.Lookup(ctx, currNode, syncedElem)
```

Comment:
It strikes me that this loop will destroy the effectiveness of the in-memory cache, since it'll keep thrashing the synced blocks into the memory cache. It'll also be doing constant reading from the disk. Not sure if we care all that much about that part.

Reply:
I don't think we can avoid the "constant reading" part given that we have to do mark-and-sweep.  Good point about the in-memory cache getting thrashed.  Do you think it's worth building some mechanism to avoid memory-caching during this process?  I lean towards no but I could be convinced.

For what it's worth, the `blockTree` is smart enough to avoid fetching any direct leaf blocks in this process, so for example the block for a small file in a directory won't be read into the cache when it's marked. (Same with all the leaf nodes of big files.)

Reply:
Some alternatives:
1. Get rid of the in-memory transient cache completely, or make it tiny (<=10MB). It won't be valuable once we have a mark and sweep trashing it.
2. Avoid caching from mark-and-sweep.
3. Have a separate cache for marking and sweeping (dirty cache, clean transient cache, and mark-sweep cache).

I'm inclined to go with 1, as the working set disk cache gives us most of the benefit of the clean transient LRU cache.

Reply:
I'm ok with 1, especially after some performance tests, but do you think it belongs in this PR?  Seems pretty unrelated.

If you want to do 2 as part of this PR, maybe we can just hack in a context value for it.

Either way, let me know what you think this PR should block on.